# 03 — Player Similarity: K-Means Clustering + PCA

Clusters PL 2015/16 outfield players into play-style archetypes, using the per-90 event metrics
built in Session S5 (`src/similarity.py`).

**Clustering is done separately per position group** (Defender / Midfielder / Forward), not all
together. Clustering everyone at once would mostly just rediscover position itself (defenders
tackle a lot, forwards shoot a lot) rather than finding meaningful sub-styles *within* a position
— which is the actually useful output for a recruitment tool ("find another presser like this one,
among forwards").

Unlike the xG model (S2-S4), there's no `is_goal` label to check a cluster against. "Is this a good
clustering" is judged by: does inertia stop dropping sharply past some K (elbow method), and —
more importantly for football — do the resulting groups read as sensible playing styles to someone
who knows the league.

**Features: 11 per-90 metrics** — non-penalty goals, shots, key passes, assists, progressive
passes, dribbles completed, pressures, interceptions, tackles, clearances, blocks. Clearances and
blocks joined on 2026-07-05 (defender-facing stats for the app). That change re-shaped the
clusters, so this notebook's narrative was re-checked against the 11-feature output on
2026-10-01 — see the Michail Antonio note below for the most interesting consequence.

In [1]:
import sys
sys.path.append('..')

import matplotlib.pyplot as plt
import pandas as pd

from src.similarity import (
    PER90_FEATURE_COLUMNS, scale_features, compute_elbow_scores, fit_kmeans,
    profile_clusters, run_pca,
)
from src.visualisation import plot_elbow_curve, plot_pca_clusters

pd.set_option('display.max_columns', 60)

features = pd.read_pickle('../data/player_per90_pl_2015_16.pkl')
features['position_group'].value_counts()

position_group
Defender      119
Midfielder    106
Forward        75
Name: count, dtype: int64

### Phase 2 rigor — minutes-weighted position assignment

The `position_group` above is now assigned by **total season minutes**, not the modal
per-match position used in S6. Modal-by-match has a real failure mode for versatile players:
when attacking minutes are split across several labels (Right Wing / Left Wing / Centre
Forward) while occasional defensive cameos all share one label, that single defensive label
can win a per-*match* vote even though attacking minutes dominate the season.
`resolve_season_positions` (in `src/similarity.py`) fixes this by totalling minutes per
*group* first, then taking the argmax.

The change reclassified **10 genuinely borderline players** — exactly the winger / forward /
midfield hybrids where "where did he log the most minutes" diverges from "what was his most
frequent match slot": Coutinho, Lingard, Mata and Sissoko move Forward → Midfielder; Firmino
and Berahino move Midfielder → Forward; Schlupp moves Midfielder → Defender; etc. Group counts
shift from 118 / 104 / 78 in S6 to **119 Defenders / 106 Midfielders / 75 Forwards**.

## Elbow method, per position group

Inertia always decreases as K grows — looking for where it stops dropping sharply, not a single
"correct" point.

In [2]:
POSITION_GROUPS = ['Defender', 'Midfielder', 'Forward']
CHOSEN_K = {'Defender': 4, 'Midfielder': 4, 'Forward': 4}

groups = {}
for position_group in POSITION_GROUPS:
    subset = features[features['position_group'] == position_group].reset_index(drop=True)
    X_scaled, scaler = scale_features(subset)
    groups[position_group] = {'data': subset, 'X_scaled': X_scaled}

fig, axes = plt.subplots(1, 3, figsize=(15, 4))
for ax, position_group in zip(axes, POSITION_GROUPS):
    inertias = compute_elbow_scores(groups[position_group]['X_scaled'], k_range=range(2, 9))
    groups[position_group]['inertias'] = inertias
    plot_elbow_curve(inertias, chosen_k=CHOSEN_K[position_group], ax=ax,
                      title=f'{position_group} (n={len(groups[position_group]["data"])})')
fig.tight_layout()
fig.savefig('../outputs/similarity_elbow_curves.png', dpi=150, bbox_inches='tight')

## Silhouette score — a quantitative second opinion on K (Phase 2 rigor)

The elbow curves above read inertia, which **always** falls as K rises — so there's no internal
optimum, you just eyeball where the bend is. The silhouette score is the complementary tool: for
each player it compares the mean distance to its *own* cluster (cohesion, `a`) against the mean
distance to the nearest *other* cluster (separation, `b`), as `(b − a) / max(a, b)`, averaged
over everyone. It ranges −1 to 1 (higher = tighter, better-separated clusters) and, unlike
inertia, does **not** improve monotonically with K — so it can be *maximised* to actually suggest
a K, rather than read by eye.

In [3]:
from src.similarity import compute_silhouette_scores
from src.visualisation import plot_silhouette_curve

fig, axes = plt.subplots(1, 3, figsize=(15, 4))
for ax, position_group in zip(axes, POSITION_GROUPS):
    silhouettes = compute_silhouette_scores(groups[position_group]['X_scaled'], k_range=range(2, 9))
    groups[position_group]['silhouettes'] = silhouettes
    plot_silhouette_curve(silhouettes, ax=ax,
                          title=f'{position_group} (n={len(groups[position_group]["data"])})')
fig.tight_layout()
fig.savefig('../outputs/similarity_silhouette_curves.png', dpi=150, bbox_inches='tight')

# Peak silhouette K and its (low) value per group — the quantitative read on K.
pd.DataFrame({
    pg: {'best_K': int(groups[pg]['silhouettes'].idxmax()),
         'max_silhouette': round(groups[pg]['silhouettes'].max(), 3)}
    for pg in POSITION_GROUPS
}).T

,best_K,max_silhouette
Defender,2.0,0.223
Midfielder,2.0,0.256
Forward,2.0,0.244


## Fit K-means (K=4 for every group) — and what silhouette says about that

The silhouette table above lands on the same answer for all three groups: **K=2, with a low
peak (~0.22–0.26)**. Two things to read from that:

1. **The absolute value is low.** A silhouette near 0.25 means the clusters overlap heavily —
   players sit much closer to their cluster's *edge* than its core. That's the honest signal
   that play-styles *within* a position are a **continuum**, not a handful of cleanly separated
   types. No K produces crisp clusters here, and pretending otherwise would be exactly the
   clean-narrative smoothing this project tries to avoid.
2. **K=2 is too coarse to be useful.** Two clusters per position collapses to roughly "more
   attacking vs. more defensive" — fine for a metric that rewards blob separation, useless as a
   recruitment-archetype lens.

So K=4 is kept **deliberately, against the metric's preferred K**: the goal is granular,
human-readable archetypes (the destroyer, the playmaker, the poacher, the presser), and because
the silhouette is low at *every* K, the cost of choosing 4 over 2 is small — we trade ~0.07–0.09 of a
silhouette point (K=4 scores ~0.15–0.17) for archetypes a scout can actually name. The metric *informs* the choice;
football usefulness and the continuum finding settle it.

`profile_clusters` then gives each cluster's average feature value as a z-score vs. the position
group's own population — the substitute for an accuracy metric, since there's no ground truth here.

In [4]:
for position_group in POSITION_GROUPS:
    g = groups[position_group]
    kmeans, labels = fit_kmeans(g['X_scaled'], n_clusters=CHOSEN_K[position_group])
    g['kmeans'] = kmeans
    g['labels'] = labels
    g['data']['cluster'] = labels
    g['profile'] = profile_clusters(g['data'], PER90_FEATURE_COLUMNS, labels)

print('Forward cluster profiles (z-scores vs. all PL 2015/16 forwards):')
groups['Forward']['profile'].round(2)

Forward cluster profiles (z-scores vs. all PL 2015/16 forwards):


,non_penalty_goals_p90,shots_p90,key_passes_p90,assists_p90,progressive_passes_p90,dribbles_completed_p90,pressures_p90,interceptions_p90,tackles_p90,clearances_p90,blocks_p90
cluster,,,,,,,,,,,
0,-0.72,-0.46,1.19,0.92,1.27,0.64,-0.30,0.51,0.03,-0.73,-0.25
1,0.92,1.00,-0.49,-0.25,-0.51,-0.50,-0.87,-0.75,-0.75,0.03,-0.80
2,-0.22,-0.21,0.47,-0.05,0.29,1.11,1.54,1.34,1.78,-0.44,1.63
3,-0.14,-0.36,-0.54,-0.36,-0.51,-0.43,0.25,-0.26,-0.12,0.58,0.14


In [5]:
print('Midfielder cluster profiles:')
display(groups['Midfielder']['profile'].round(2))
print('Defender cluster profiles:')
display(groups['Defender']['profile'].round(2))

Midfielder cluster profiles:


,non_penalty_goals_p90,shots_p90,key_passes_p90,assists_p90,progressive_passes_p90,dribbles_completed_p90,pressures_p90,interceptions_p90,tackles_p90,clearances_p90,blocks_p90
cluster,,,,,,,,,,,
0,-0.72,-0.86,-0.81,-0.67,0.37,-0.54,0.30,0.81,0.72,0.93,0.50
1,0.17,0.28,0.23,0.20,-0.17,-0.24,0.45,-0.23,0.14,-0.24,0.26
2,0.18,0.47,0.31,0.03,-0.20,0.74,-1.12,-0.76,-1.01,-0.79,-1.09
3,1.72,1.34,1.62,1.80,-0.43,1.01,0.06,-0.63,-0.78,-1.00,-0.16


Defender cluster profiles:


,non_penalty_goals_p90,shots_p90,key_passes_p90,assists_p90,progressive_passes_p90,dribbles_completed_p90,pressures_p90,interceptions_p90,tackles_p90,clearances_p90,blocks_p90
cluster,,,,,,,,,,,
0,-0.11,-0.34,-0.66,-0.52,-0.59,-0.47,-0.59,-0.32,-0.63,0.55,-0.16
1,-0.41,-0.38,0.46,0.23,0.15,0.04,0.98,0.22,0.91,-0.58,0.53
2,0.57,0.13,-0.64,-0.62,1.09,-0.67,-0.87,0.60,-0.52,1.02,-0.57
3,0.39,1.31,1.20,1.34,-0.32,1.69,0.29,-0.46,0.12,-1.18,-0.06


## PCA scatter — 2D view of each position group's clusters

PCA compresses the 11 per-90 features down to 2 components for plotting. Each axis is a blend of
the original features (not a single metric), so this trades away individual-feature
interpretability in exchange for being able to see cluster separation at all — the
`explained_variance_ratio_` below shows how much of the original signal survives that compression.

In [6]:
fig, axes = plt.subplots(1, 3, figsize=(16, 5))
for ax, position_group in zip(axes, POSITION_GROUPS):
    g = groups[position_group]
    components, pca = run_pca(g['X_scaled'])
    g['pca_components'] = components
    g['pca'] = pca
    plot_pca_clusters(components, g['labels'], ax=ax,
                       title=f'{position_group}\n(explained var: {pca.explained_variance_ratio_.sum():.0%})')
fig.tight_layout()
fig.savefig('../outputs/similarity_pca_clusters.png', dpi=150, bbox_inches='tight')

## Reading the archetypes — who's actually in each cluster

The real test of whether this clustering is useful: do the players grouped together actually play
similar roles? Showing the highest-minutes players per cluster (most data behind their per-90
rates, so the least noisy examples).

In [7]:
for position_group in POSITION_GROUPS:
    data = groups[position_group]['data']
    print(f'=== {position_group} ===')
    for cluster_id in sorted(data['cluster'].unique()):
        top_players = (
            data[data['cluster'] == cluster_id]
            .sort_values('minutes_played', ascending=False)
            .head(5)['player'].tolist()
        )
        print(f'  Cluster {cluster_id} (n={sum(data["cluster"] == cluster_id)}): {", ".join(top_players)}')
    print()

=== Defender ===
  Cluster 0 (n=40): Wes Morgan, Craig Dawson, Robert Huth, Scott Dann, Chris Smalling
  Cluster 1 (n=38): Simon Francis, Ignacio Monreal Eraso, César Azpilicueta Tanco, Daley Blind, Erik Pieters
  Cluster 2 (n=22): Toby Alderweireld, Ashley Williams, Steve Cook, José Miguel da Rocha Fonte, Virgil van Dijk
  Cluster 3 (n=19): Aaron Cresswell, Charlie Daniels, Héctor Bellerín Moruno, Patrick van Aanholt, Robert Brady

=== Midfielder ===
  Cluster 0 (n=40): Andrew Surman, Eric Dier, Glenn Whelan, Yann Gérard M''Vila, Idrissa Gana Gueye
  Cluster 1 (n=31): Mark Noble, Danny Drinkwater, Matt Ritchie, Jonathan Howson, Francesc Fàbregas i Soler
  Cluster 2 (n=24): Darren Fletcher, Moussa Sissoko, Georginio Wijnaldum, Ross Barkley, Juan Manuel Mata García
  Cluster 3 (n=11): Mesut Özil, Riyad Mahrez, Gylfi Þór Sigurðsson, Sadio Mané, Bamidele Alli

=== Forward ===
  Cluster 0 (n=17): Christian Dannemann Eriksen, Marko Arnautović, Willian Borges da Silva, Dimitri Payet, Alexis 

**Michail Antonio — the outlier, and what the cluster structure does (and doesn't) show.**
Checking his real 2015/16 minutes (Phase 2) disproved the S6 assumption that he was a mislabelled
winger: they split **920 min at right-back / wing-back, 761 min as a winger, 452 in midfield**.
The *plurality* is genuinely defensive, so the minutes-weighted rule correctly keeps him a
Defender — the fix did its job elsewhere (10 sensible reclassifications) without touching him.

He is still, by far, the most extreme defender in the data: z-scores of **+6.0 non-penalty
goals/90, +5.1 shots/90, +4.2 dribbles/90** against other defenders, and his nearest neighbour
(Patrick van Aanholt) sits **6.3** standardised units away — 2–3× a typical "players like X"
match (Kanté's closest is 2.8, Cresswell's 1.9, see below). With the original 9 features, K-means
spent one of its four defender clusters on him alone. **With 11 features it doesn't:** the fourth
cluster now captures a real archetype — **ball-playing centre-backs** (Alderweireld, van Dijk,
Koscielny: high progressive passes *and* clearances) — and Antonio is absorbed into the attacking
full-back cluster as its farthest member (7.7 units from its centroid; the next-farthest, 4.7).

That's K-means' objective at work, not a change in the player: a singleton cluster survives only
while one point's squared distance costs more than splitting a genuine group, and the two new
defensive features created a split worth more. **The lesson: a hard cluster label can hide an
outlier as easily as it can reveal one.** The distances and the radar still show Antonio as a true
positional hybrid (forward-level output from full-back minutes); the label alone no longer does.
Multi-position / soft membership (the Phase 6c GMM idea) remains the principled fix for a player
who legitimately belongs partly to two groups at once.

## S6 takeaways

- Clustering per position group (rather than all outfield players together) was the right call —
  it surfaces play-style differences *within* a position instead of just rediscovering position
  itself.
- No accuracy metric exists for this the way ROC-AUC exists for the xG model — judgment here comes
  from reading whether the cluster profiles and player groupings make football sense, which they
  do (Kane/Vardy/Agüero/Giroud/Lukaku/Defoe share the penalty-box-finisher forward cluster;
  Kanté/Gueye/M'Vila/Tioté the ball-winning midfield one; Cresswell/Bellerín/Daniels/van Aanholt
  the attacking full-back one).
- PCA's `explained_variance_ratio_` quantifies what the 2D scatter actually shows vs. hides — worth
  reading before trusting a visual cluster gap or overlap at face value.
- Next session (S7): radar charts per player (these per-90 features, presented per-position), PCA
  scatter as a standalone deliverable, and a "players like X" lookup function built on top of these
  same clusters/distances.

# Session S7 — Radar Charts + "Players Like X"

Two deliverables this session, both built directly on the per-90 features and clusters from S5/S6:

1. **Radar charts** (`plot_player_radar`, `src/visualisation.py`) — a per-player visual profile
   across the same 11 per-90 metrics used for clustering, scaled to their position group's 5th-95th
   percentile range (not true min/max — see the function's docstring for why: a true-range radar
   would let the Michail Antonio outlier (above) compress every other defender's chart into an
   unreadable sliver).
2. **"Players like X" lookup** (`find_similar_players`, `src/similarity.py`) — nearest neighbours
   by Euclidean distance in standardised feature space, within the same position group. This is
   deliberately *not* "same K-means cluster as X": cluster membership is a coarse in/out split,
   while raw distance gives a continuous, rankable "most to least similar" — the more useful shape
   for an actual scouting lookup tool.

The PCA scatter plot (the third S7 deliverable per the original roadmap) was already built and
validated in S6 above (`plot_pca_clusters`) — nothing new needed there, it's already a standalone
deliverable in its own right.

In [8]:
from src.similarity import find_similar_players
from src.visualisation import plot_player_radar

## Radar charts — one example player per position group

Picking one recognisable player from each position group's archetypes above: N'Golo Kanté
(midfielder, ball-winning destroyer cluster — stored in this dataset as `N''Golo Kanté` with a
doubled apostrophe, a StatsBomb data quirk rather than a typo here), Aaron Cresswell (defender,
attacking full-back cluster), and Harry Kane (forward, pure-finisher cluster) — each radar is
scaled to their own position group's 5th-95th percentile range, so "near the edge" means genuinely
unusual for that position, not just a high raw number.

In [9]:
radar_examples = [
    ('Midfielder', "N''Golo Kanté", 'Leicester City'),
    ('Defender', 'Aaron Cresswell', 'West Ham United'),
    ('Forward', 'Harry Kane', 'Tottenham Hotspur'),
]

fig, axes = plt.subplots(1, 3, figsize=(18, 7))
for ax, (position_group, player, team) in zip(axes, radar_examples):
    data = groups[position_group]['data']
    row = data[(data['player'] == player) & (data['team'] == team)].iloc[0]
    plot_player_radar(
        row, population=data, feature_columns=PER90_FEATURE_COLUMNS, ax=ax,
        title=f'{player}\n{position_group} — {team}',
    )
fig.tight_layout()
fig.savefig('../outputs/player_radar_examples.png', dpi=150, bbox_inches='tight')

## "Players like X" — nearest-neighbour lookup

Same three players as above, now asking "who else plays like this" rather than "what does this
player's profile look like". `find_similar_players` ranks every other player in the same position
group by Euclidean distance in standardised feature space — the same scaled feature matrix the
clustering itself uses, just read as a continuous ranking instead of a discrete cluster label.

In [10]:
for position_group, player, team in radar_examples:
    print(f'Players like {player} ({position_group}):')
    similar = find_similar_players(
        groups[position_group]['data'], PER90_FEATURE_COLUMNS, player, team, n=5,
    )
    display(similar[['player', 'team', 'distance']].round(2))
    print()

Players like N''Golo Kanté (Midfielder):


,player,team,distance
0,Cheik Ismaël Tioté,Newcastle United,2.79
1,Idrissa Gana Gueye,Aston Villa,2.88
2,Jan Kirchhoff,Sunderland,3.21
3,Claudio Ariel Yacob,West Bromwich Albion,3.23
4,Jack Frank Porteous Cork,Swansea City,3.28



Players like Aaron Cresswell (Defender):


,player,team,distance
0,Chris Brunt,West Bromwich Albion,1.92
1,Ben Davies,Tottenham Hotspur,2.30
2,Bacary Sagna,Manchester City,2.42
3,Ryan Bertrand,Southampton,2.56
4,Kyle Naughton,Swansea City,2.71



Players like Harry Kane (Forward):


,player,team,distance
0,Odion Jude Ighalo,Watford,2.20
1,Jamie Vardy,Leicester City,2.31
2,Olivier Giroud,Arsenal,2.57
3,Diego da Silva Costa,Chelsea,2.62
4,Theo Walcott,Arsenal,2.62


## S7 takeaways

- Radar charts use percentile-bounded axes (5th-95th, not true min/max) specifically because of
  the S6 Antonio outlier — a true-range axis would compress every other defender's chart toward
  the centre on every feature, not just the one he's actually unusual in.
- "Players like X" deliberately ranks by raw distance rather than just returning same-cluster
  players — a continuous ranking is more useful for a scouting lookup than a coarse in/out label,
  and the two aren't the same information (same cluster doesn't mean equally similar).
- Sanity check passed cleanly: nearest neighbours for Kanté, Cresswell, and Kane were all
  recognisable same-role players, not noise — see the lookup output above.
- Hit a real StatsBomb data quirk along the way: this dataset stores Kanté's name as
  `N''Golo Kanté` (doubled apostrophe), not the more familiar single-apostrophe spelling — noted
  in `ML_LEARNING_LOG.md` so it isn't re-discovered as a bug later.
- Next session (S8): README narrative, outputs committed, repo polish for CV/LinkedIn links.